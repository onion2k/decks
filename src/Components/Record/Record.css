
.Deck {
    position: absolute;
    top: 0;
    width: 950px;
    height: 950px;
    right: -400px;
    z-index: 0;
}

.Record {
    font-family: 'Roboto Condensed', sans-serif;
    position: absolute;
    top: 0;
    width: 950px;
    height: 950px;
    right: 0;
    transform: translate3d(0, 0, 0);
}

.DiscTwo {
    right: 100px;
}

.title {
    position: absolute;
    bottom: 0;
    left: 50%;
    width: 280px;
    height: 120px;
    padding: 20px 100px;
    margin-left: -240px;
    text-align: center;
    background-color: #FFA500;
    color: #484848;
}

.label {
    position: absolute;
    width: 480px;
    height: 480px;
    top: 50%;
    left: 50%;
    margin-left: -240px;
    margin-top: -240px;
    clip-path: circle(200px at center);
    transition: opacity 2s ease-in;
}

.Record.playing {
    animation: spin 8.33s infinite linear;
}

@keyframes spin {
    from { transform:rotate(0deg); }
    to { transform:rotate(360deg); }
}

.Tonearm {
    position: absolute;
    top: 0px;
    left: -500px;
    width: 950px;
    height: 950px;
    z-index: 1;
    transform: scaleX(-1) scaleY(-1);
    user-drag: none; 
    user-select: none;
    -moz-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
    -ms-user-select: none;
}

.Tonearm > img {
    transform-origin: 372px 77px;
    transform:rotate(-54deg);
}

.Tonearm.playing > img {
    /* animation: swing 1080s infinite linear; */
}

@keyframes swing {
    0% { transform:rotate(-49deg); }
    99% { transform:rotate(-20deg); }
    100% { transform:rotate(-49deg); }
}