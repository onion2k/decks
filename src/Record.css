@import url('https://fonts.googleapis.com/css?family=Roboto+Condensed');

.Record {
    font-family: 'Roboto Condensed', sans-serif;
    position: absolute;
    top: 0;
    width: 950px;
    height: 950px;
    right: -20%;
    transform: translate3d(0, 0, 0);
    z-index: -1;
}

.title {
    position: absolute;
    bottom: 0;
    left: 50%;
    width: 280px;
    height: 120px;
    padding: 20px 100px;
    margin-left: -240px;
    text-align: center;
    background-color: #FFA500;
    color: #484848;
}

.label {
    position: absolute;
    width: 480px;
    height: 480px;
    top: 50%;
    left: 50%;
    margin-left: -240px;
    margin-top: -240px;
    clip-path: circle(200px at center);
    transition: opacity 2s ease-in;
}

.spinning {
    animation: spin 8.33s infinite linear;
}

@keyframes spin {
    from { transform:rotate(0deg); }
    to { transform:rotate(360deg); }
}
