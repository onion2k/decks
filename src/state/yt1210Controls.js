import { configure, decorate, observable, action } from "mobx";

import fx from "./yt1210Sounds";

configure({ enforceActions: true });

export default class yt1210Controls {

    constructor(pm){
      this.pm = pm;
    }
    
    videoId = null;
  
    playing = false;
    repeat = false;
    shuffle = false;
    crackle = true;
    video = true;
  
    // play track
    play = (id = null) => {
      if (id!==null) {
        this.videoId = this.pm.getCurrentTrack().videoId;
      } else {
        this.videoId = this.pm.getTrackById(id).videoId;
      }
      this.playing = true;
    };
  
    // pause track
    pause = () => {
      this.playing = !this.playing;
    };
  
    // stop track
    stop = () => {
      this.playing = false;
    };
  
    // next track
    next = () => {
      this.videoId = this.pm.getNextTrack().videoId;
      this.playing = true;
    };
  
    // copy playlist from youtube

    toggle_repeat = () => {
      this.repeat = !this.repeat;
      if (this.repeat) { fx.ogg_scratchin.play(); }
    }
  
    toggle_shuffle = () => {
      this.shuffle = !this.shuffle;
    };
  
    toggle_crackle = () => {
      this.crackle = !this.crackle;
    };

    toggle_video = () => {
      this.video = !this.video;
    };
  }
  
  decorate(yt1210Controls, {
    play: action,
    pause: action,
    stop: action,
    next: action,
    toggle_repeat: action,
    toggle_shuffle: action,
    toggle_crackle: action,
    toggle_video: action,
    videoId: observable,
    playing: observable,
    crackle: observable,
    video: observable,
    repeat: observable,
    shuffle: observable
  });
  
