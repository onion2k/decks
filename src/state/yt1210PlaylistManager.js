import { configure, decorate, observable, action, computed } from "mobx";

configure({ enforceActions: true });

export default class yt12010PlaylistManager {

    // Playlists

    playlist = 1;
    track = 0;
    currentTrack = { videoId: "", label: "" };

    playlists = [
        { "id": 1, "title": "Playlist 1", tracks: [
          { title: "", videoId: "BuVJEn9wk9Y", duration: "" },
          { title: "", videoId: "Rs38lKxmtI4", duration: "" },
          { title: "", videoId: "v_yTphvyiPU", duration: "" },
          { title: "", videoId: "e4Ao-iNPPUc", duration: "" },
          { title: "", videoId: "FALYmqt-7TQ", duration: "" },
          { title: "", videoId: "a3Z4RWZa9WA", duration: "" }
        ]},
        { "id": 2, "title": "Playlist 2", tracks: [
            { title: "", videoId: "a3Z4RWZa9WA", duration: "" },
            { title: "", videoId: "BuVJEn9wk9Y", duration: "" },
            { title: "", videoId: "Rs38lKxmtI4", duration: "" },
            { title: "", videoId: "v_yTphvyiPU", duration: "" },
            { title: "", videoId: "e4Ao-iNPPUc", duration: "" },
            { title: "", videoId: "FALYmqt-7TQ", duration: "" }
          ]}
    ];

    load = (id) => {
        this.playlist = id;
        // const pl = this.playlists.find((playlist) => playlist.id === this.playlist).tracks;
    };

    // add playlist
    playlistAdd = (title)=>{
        this.playlists.push({ id: this.playlists.length+1, title: title, tracks: [] });
    };

    // delete playlist
    playlistDelete = (index)=>{
        this.playlists.splice(index, 1);
    };

    // add track
    playlistAddTrack = (id, videoId)=>{
        this.playlist = this.playlists.find((playlist) => playlist.id === id);
        this.playlist.tracks.push({ title: "", videoId: videoId, duration: "" });
    };

    // delete track
    playlistDeleteTrack = (id, index)=>{
        this.playlist = this.playlists.find((playlist) => playlist.id === id);
        this.playlist.tracks.splice(index, 1);
    };

    get tracks() {
        return this.playlists.find((playlist) => playlist.id === this.playlist).tracks;
    };

    updateTrackData = (data, duration) => {
        const track = this.tracks.find((track)=>track.videoId === data.video_id);
              track.title = data.title;
              track.duration = duration;
        this.currentTrack = track;
    };

    getTracks = () => {
        return this.tracks;
    };

    getCurrentTrack = () => {
        return this.tracks[this.track];
    };

    getNextTrack = () => {
        return this.tracks[this.track++];
    };

    getTrackById = (id) => {
        this.track = this.tracks.findIndex((track) => track.videoId === id);
        return this.tracks[this.track];
    };
}

decorate(yt12010PlaylistManager, {
    currentTrack: observable,
    playlists: observable,
    playlist: observable,
    track: observable,
    tracks: computed,
    load: action,
    playlistAdd: action,
    playlistDelete: action,
    playlistAddTrack: action,
    playlistDeleteTrack: action,
    updateTrackData: action,
    getCurrentTrack: action,
    getNextTrack: action
});
