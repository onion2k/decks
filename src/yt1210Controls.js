// play
// stop
// pause
// next
// previous

// tonearm move
