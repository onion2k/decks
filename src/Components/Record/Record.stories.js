import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Record from '../Components/Record';

storiesOf('Record', module)
  .add('Basic', () => (
    <Record />
  ));
