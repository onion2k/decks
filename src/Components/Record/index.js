import React, { Component } from "react";
import { inject, observer } from "mobx-react";
import { autorun } from "mobx";
import record from "../../img/record.png";
import tonearm from "../../img/tonearm.png";
import "./Record.css";

const tracks = [
  { s: -47.6, e: -43.4 },
  { s: -43.4, e: -39.7 },
  { s: -39.7, e: -34.2 },
  { s: -34.2, e: -28.5 },
  { s: -28.5, e: -25.6 },
  { s: -25.6, e: -20.0 }
];

class Record extends Component {

  constructor(props) {
    super(props);

    // this.armMove = this.armMove.bind(this);
    // this.armUp = this.armUp.bind(this);
    // this.tick = this.tick.bind(this);
    this.lastX = 0;

    this.requestAnimationFrameId = null;

    // this.state = {
    //   requestAnimationFrameId: null,
    //   title: "Unknown",
    //   label: "",
    //   dragtone: false,
    //   tonestart: -54,
    //   toneangle: -54,
    //   tonestyle: "rotate(-54deg)",
    //   millis: Date.now()
    // };
  }

  track = autorun(() => {
    const { track } = this.props.playlistManager;

    if (track) {
      this.setState({
        toneanimto: tracks[track-1].s
      });
      // if (this.props.playlistControls.playing === true) {
      //   if (!this.requestAnimationFrameId) {
      //     this.requestAnimationFrameId = requestAnimationFrame(this.tick);
      //   }
      // } else if (this.requestAnimationFrameId) {
      //   cancelAnimationFrame(this.requestAnimationFrameId);
      // }
    }

  });

  // tick() {
  //   var toneangle, toneanimto;

  //   toneanimto = this.state.toneanimto;

  //   if (this.state.toneanimto === null) {
  //     toneanimto = null;
  //     toneangle = this.state.toneangle + 0.001; // This will die on it's arse on a long track
  //   } else {
  //     if (this.state.toneanimto > this.state.toneangle) {
  //       toneangle = this.state.toneangle + 0.15;
  //     } else {
  //       toneangle = this.state.toneangle - 0.15;
  //     }

  //     if (Math.abs(toneangle - this.state.toneanimto) < 0.2) {
  //       toneanimto = null;
  //     }
  //   }

    // this.setState({
    //     millis: Date.now(),
    //     requestAnimationFrameId: requestAnimationFrame(this.tick),
    //     tonestart: toneangle,
    //     toneangle: toneangle,
    //     tonestyle: 'rotate('+toneangle+'deg)',
    //     toneanimto: toneanimto
    // });
  //  }

  // armMove(e) {
  //   if (this.state.dragtone) {
  //     if (this.state.requestAnimationFrameId) {
  //       cancelAnimationFrame(this.state.requestAnimationFrameId);
  //     }
  //     let toneangle = this.state.tonestart + (e.pageX - this.lastX) / 10;
  //     if (toneangle < -54) {
  //       toneangle = -54;
  //     }
  //     if (toneangle > -19) {
  //       toneangle = -19;
  //     }
  //     this.setState({
  //       requestAnimationFrameId: false,
  //       toneangle: toneangle,
  //       tonestyle: "rotate(" + toneangle + "deg)"
  //     });
  //   }
  // }

  // armUp(e) {
  //   var toneangle, per;

  //   let perCal = function(s, e, p) {
  //     let tl = Math.abs(s) - Math.abs(e);
  //     let per = p + Math.abs(s);
  //     return (per / tl) * 100;
  //   };

  //   this.setState({ dragtone: false, tonestart: this.state.toneangle });

  //   toneangle = this.state.toneangle;

  //   if (toneangle > -20) {
  //     //off
  //   } else if (toneangle > -25.6) {
  //     per = perCal(-25.6, -20, toneangle);
  //     this.props.onTrack(5, per);
  //   } else if (toneangle > -28.5) {
  //     per = perCal(-28.5, -25.6, toneangle);
  //     this.props.onTrack(4, per);
  //   } else if (toneangle > -34.2) {
  //     per = perCal(-34.2, -28.5, toneangle);
  //     this.props.onTrack(3, per);
  //   } else if (toneangle > -39.7) {
  //     per = perCal(-39.7, -34.2, toneangle);
  //     this.props.onTrack(2, per);
  //   } else if (toneangle > -43.4) {
  //     per = perCal(-43.4, -39.7, toneangle);
  //     this.props.onTrack(1, per);
  //   } else if (toneangle > -47.6) {
  //     per = perCal(-47.6, -43.4, toneangle);
  //     this.props.onTrack(0, per);
  //   } else {
  //     //off
  //   }
  // }

  render() {

    return (
      <div
        className="Deck"
        // onMouseDown={e => {
        //   this.lastX = e.pageX;
        //   this.setState({ dragtone: true });
        // }}
        // onMouseUp={e => {
        //   this.armUp(e);
        // }}
        // onMouseMove={e => {
        //   this.armMove(e);
        // }}
      >
        <div
          className={
            "Record" + (this.props.playlistControls.playing === true ? " playing" : "")
          }
          style={{ backgroundImage: "url(" + record + ")" }}
        >
          <div className="label">
            <img src={"https://img.youtube.com/vi/" + this.props.playlistManager.currentTrack.videoId + "/0.jpg"} alt="Record" />
            <span className="title">{this.props.playlistManager.currentTrack.title}</span>
          </div>
        </div>
        <div
          className={
            "Tonearm" + (this.props.playlistControls.playing === true ? " playing" : "")
          }
        >
          <img
            draggable="false"
            src={tonearm}
            alt="Tonearm"
            style={{ transform: "" }}
          />
        </div>
      </div>
    );
  }
}

export default inject("playlistControls", "playlistManager")(observer(Record));
