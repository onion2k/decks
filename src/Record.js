import React, { Component } from 'react';
import record from './img/record.png';
import tonearm from './img/tonearm.png';
import './Record.css';

class Record extends Component {

    constructor(props) {
        super(props);
        this.state = { title: 'Unknown', label: '' }
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            label: 'https://img.youtube.com/vi/'+nextProps.vData.videoId+'/0.jpg',
            title: nextProps.vData.title
        })
    }

    render() {
        return (
            <div className="Deck">
                <div className={'Record'+(this.props.playing===true?' playing':'')} style={{ backgroundImage: 'url('+record+')' }} >
                    <div className="label">
                        <img src={ this.state.label } alt='Record' />
                        <span className="title">{ this.state.title }</span>
                    </div>
                </div>
                <div className={'Tonearm'+(this.props.playing===true?' playing':'')}>
                    <img src={ tonearm } alt='Tonearm' />
                </div>
            </div>
        );
    }
}

export default Record;
