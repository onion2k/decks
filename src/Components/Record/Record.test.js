import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from 'react-router-dom'

import Record from ".";

import yt1210PlaylistManager from "../../state/yt1210PlaylistManager"
const playlistManagerState = new yt1210PlaylistManager();

import yt1210Controls from "../../state/yt1210Controls"
const controlState = new yt1210Controls();

describe("Record", () => {
  it("renders properly", () => {
    const tree = renderer.create(
    <Record playlistManager={playlistManagerState} playlistControls={controlState} />
  ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
