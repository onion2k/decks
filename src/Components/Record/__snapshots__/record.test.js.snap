// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Record renders properly 1`] = `
<div
  className="Deck"
  onMouseDown={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="Record"
    style={
      Object {
        "backgroundImage": "url(record.png)",
      }
    }
  >
    <div
      className="label"
    >
      <img
        alt="Record"
        src="https://img.youtube.com/vi//0.jpg"
      />
      <span
        className="title"
      />
    </div>
  </div>
  <div
    className="Tonearm"
  >
    <img
      alt="Tonearm"
      draggable="false"
      src="tonearm.png"
      style={
        Object {
          "transform": "rotate(-54deg)",
        }
      }
    />
  </div>
</div>
`;
